packages/expo-screen-orientation/ios/ScreenOrientationRegistry.swift

```diff
@@ -0,0 +1,239 @@
+    if currentOrientationMask.contains(newScreenOrientation) {
+      // when changing orientation without changing dimensions traitCollectionDidChange isn't triggered so the event has to be called manually
+      if (newScreenOrientation.isPortrait && currentScreenOrientation.isPortrait)
+        || (newScreenOrientation.isLandscape && currentScreenOrientation.isLandscape) {
+        screenOrientationDidChange(newScreenOrientation)
+        return
+      }
+
+      // on iPads, traitCollectionDidChange isn't triggered at all, so we have to call screenOrientationDidChange manually
+      if isIPad()
+        && (newScreenOrientation.isPortrait && currentScreenOrientation.isLandscape
+        || newScreenOrientation.isLandscape && currentScreenOrientation.isPortrait) {
+        screenOrientationDidChange(newScreenOrientation)
+      }
+    }
+  }
+
+  // Called by ScreenOrientationViewController when the dimensions of the view change
+  // Also used for Expo Go in EXAppViewController
+  @objc
+  public func traitCollectionDidChange(to traitCollection: UITraitCollection) {
+    currentTraitCollection = traitCollection
+
+    let verticalSizeClass = traitCollection.verticalSizeClass
+    let horizontalSizeClass = traitCollection.horizontalSizeClass
+    let currentDeviceOrientation = UIDevice.current.orientation.toInterfaceOrientation()
+    let currentOrientationMask = self.rootViewController?.supportedInterfaceOrientations ?? []
+
+    var newScreenOrientation = UIInterfaceOrientation.unknown
+
+    // We need to deduce what is the new screen orientaiton based on currentOrientationMask and new dimensions of the view
+    if verticalSizeClass == .regular && horizontalSizeClass == .compact {
+      // From trait collection, we know that screen is in portrait or upside down orientation.
+      let portraitMask = currentOrientationMask.intersection([.portrait, .portraitUpsideDown])
+
+      // Mask allows only proper portrait - we know that the device is in either proper portrait or upside down
+      // we deduce it is proper portrait.
+      if portraitMask == .portrait {
+        newScreenOrientation = .portrait
+      }
+      // Mask allows only upside down portrait - we know that the device is in either proper portrait or upside down
+      // we deduce it is upside down portrait.
+      else if portraitMask == .portraitUpsideDown {
+        newScreenOrientation = .portraitUpsideDown
+      }
+      // Mask allows portrait or upside down portrait - we can try to deduce orientation
+      // from device orientation.
+      else if currentDeviceOrientation == .portrait || currentDeviceOrientation == .portraitUpsideDown {
+        newScreenOrientation = currentDeviceOrientation
+      }
+    } else if (verticalSizeClass == .compact && horizontalSizeClass == .compact)
+      || (verticalSizeClass == .regular && horizontalSizeClass == .regular)
+      || (verticalSizeClass == .compact && horizontalSizeClass == .regular) {
```

Comment:
For more complex conditions like this, I would recommend extracting them to a separate function